design/src/rfcs/rfc0017_customizable_client_operations.md

```diff
@@ -0,0 +1,185 @@
+RFC: Customizable Client Operations
+===================================
+
+> Status: RFC
+
+For a summarized list of proposed changes, see the [Changes Checklist](#changes-checklist) section.
+
+SDK customers occasionally need to add additional HTTP headers to requests, and currently,
+the SDK has no easy way to accomplish this. At time of writing, the lower level Smithy
+client has to be used to create an operation, and then the HTTP request augmented on
+that operation type. For example:
+
+```rust
+let input = SomeOperation::builder().some_value(5).build()?;
+
+let operation = {
+    let op = input.make_operation(&service_config).await?;
+    let (request, response) = op.into_request_response();
+
+    let request = request.augment(|req, _props| {
+        req.headers_mut().insert(
+            HeaderName::from_static("x-some-header"),
+            HeaderValue::from_static("some-value")
+        );
+        Result::<_, Infallible>::Ok(req)
+    })?;
+
+    Operation::from_parts(request, response)
+};
+
+let response = smithy_client.call(operation).await?;
+```
+
+This approach is both difficult to discover and implement since it requires acquiring
+a Smithy client rather than the generated fluent client, and it's anything but ergonomic.
+
+This RFC proposes an easier way to augment requests that is compatible with the fluent
+client.
+
+Terminology
+-----------
+
+- **Smithy Client**: A `aws_smithy_client::Client<C, M, R>` struct that is responsible for gluing together
+  the connector, middleware, and retry policy.
+- **Fluent Client**: A code generated `Client` that has methods for each service operation on it.
+  A fluent builder is generated alongside it to make construction easier.
+
+Proposal
+--------
+
+The code generated fluent builders returned by the fluent client should have a method added to them,
+similar to `send`, but that returns a customizable request. The customer experience should look as
+follows:
+
+```rust
+let response = client.some_operation()
+    .some_value(5)
+    .customize()?
+    .map_request(|mut req| {
+        req.headers_mut().insert(
+            HeaderName::from_static("x-some-header"),
+            HeaderValue::from_static("some-value")
+        );
+        Result::<_, Infallible>::Ok(req)
+    })?
+    .send()
+    .await?;
+```
+
+This new `customize` method would return the following:
+
+```rust
+pub struct CustomizableOperation<O, R> {
+    handle: Arc<Handle>,
+    operation: Operation<O, R>,
+}
+
+impl<O, R> CustomizableOperation<O, R> {
+    // Allows for customizing the operation's request
+    fn map_request<E>(
```

Comment:
we could also add `mutate_request` for the common case where we don't need to take ownership and the changes are infallible

fn mutate_request(self, f: impl FnOnce(&mut Request<SdkBody)) -> Self { ... }

Reply:
Good idea!